### PR TITLE
telemetry: Add `sessionId` to metrics

### PR DESCRIPTION
**Pull request by nkomonen-amazon**
## Problem:

I have filtered all the metrics in Kibana by `clientId: "XXXXXX"`, but if they had multiple IDE windows opened at the same time sending metrics I cannot know which window they are coming from since they share the same clientID.

## Solution:

This adds a new field to the telemetry service  called `sessionId` which uniquely identifies a part of the UI in an IDE window. It is adjacent to `clientId`.

Now in my metrics I will be able to differentiate metrics by their sessionID since they both share the same clientId by design.

## Implementation Details:

Changes in the Telemetry service need to be merged before this change can since they update the endpoint to accept this new field.

The updates of this commit only mirror the upstream, so this process is manual.

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by hayemaxi** [comment]
How is sessionId generated and sent? Will there be code changes to generate this sessionId later, like how we generate clientId?

**Comment by nkomonen-amazon** [comment]
@hayemaxi see the new changes that actually send a sessionID

File: packages/core/src/shared/telemetry/util.ts [hunk 1]
```diff
@@ -90,6 +91,37 @@ export function convertLegacy(value: unknown): boolean {
...
+ * value looks close to a UUID by does not exactly match it (has additional characters).
+ * As a result we process it through uuidV5 which creates a proper UUID from it.
+ * uuidV5 is idempotent, so as long as `vscode.env.sessionId` returns the same value,
+ * we will get the same UUID.
+ */
+export const getSessionId = once(() => uuidV5(vscode.env.sessionId, sessionIdNonce))
+/**
+ * This is a nonce that is used in creating a v5 UUID for Session ID.
+ * - This should ONLY be used by {@link getSessionId}.
+ * - This should NEVER change
```
Comment (hayemaxi) [inline packages/core/src/shared/telemetry/util.ts:120]: Why?

File: packages/core/src/shared/telemetry/util.ts [hunk 2]
```diff
@@ -90,6 +91,37 @@ export function convertLegacy(value: unknown): boolean {
...
+ *
+ * - On startup of a new application instance, a new session ID must be created.
+ *   - This identifier must be a `UUID`, it is enforced by the telemetry service.
+ * - The session ID must be different from all other IDs on the machine
+ * - A session ID exists until the application instance is terminated.
+ *   It should never be used again after termination.
+ * - All extensions on the same application instance should return the same session ID.
+ *
+ * `vscode.env.sessionId` behaves as described aboved, EXCEPT its
+ * value looks close to a UUID by does not exactly match it (has additional characters).
```
Comment (hayemaxi) [inline packages/core/src/shared/telemetry/util.ts:111]: Why does it matter what it looks like if its a unique and random value?

File: packages/core/src/shared/crypto.ts [hunk 3]
```diff
@@ -30,3 +30,15 @@ export function randomUUID(): `${string}-${string}-${string}-${string}-${string}
...
     return require('crypto').randomUUID()
 }
+
+/**
+ * Returns true if the given string is a UUID
+ *
+ * NOTE: There are different UUID versions, this function does not discriminate between them.
+ * See: https://stackoverflow.com/questions/7905929/how-to-test-valid-uuid-guid
+ */
+export function isUUID(uuid: string): boolean {
```
Comment (hayemaxi) [inline packages/core/src/shared/crypto.ts:40]: What might use this? Besides tests

Comment (nkomonen-amazon) [reply packages/core/src/shared/crypto.ts:40]: Anything that needs to validate if a given string is in a valid UUID format

File: packages/core/src/shared/telemetry/util.ts (hunk 2 above)
Comment (nkomonen-amazon) [reply packages/core/src/shared/telemetry/util.ts:111]: The server enforces it, mentioned here: https://github.com/aws/aws-toolkit-vscode/pull/5510/files#diff-6b731a821720f9104e0c179a1bc64d1039af6d386549d43090cc5c661be0f79eR104

File: packages/core/src/shared/telemetry/util.ts (hunk 1 above)
Comment (nkomonen-amazon) [reply packages/core/src/shared/telemetry/util.ts:120]: It is an arbitrary value that is required for UUIDv5. There is no reason to change it, and if it is modified part way through a session we would get a different UUID for the same input (though we do not currently run in to this due to the use once `once()` above.

File: packages/core/src/shared/crypto.ts (hunk 3 above)
Comment (justinmk3) [inline packages/core/src/shared/crypto.ts:40]:
```suggestion
export function isUuid(uuid: string): boolean {
```

File: packages/core/src/shared/crypto.ts [hunk 4]
```diff
@@ -30,3 +30,15 @@ export function randomUUID(): `${string}-${string}-${string}-${string}-${string}
...
+    // pattern match the uuid to a uuid string
```
Comment (justinmk3) [inline packages/core/src/shared/crypto.ts:41]: looks like a normal regex, what is the purpose of this comment? 

however it's worth mentioning:

    // Must match "SessionID" pattern in telemetry/service-2.json

File: packages/core/src/test/shared/crypto.test.ts [hunk 5]
```diff
@@ -25,4 +25,22 @@ describe('crypto', function () {
             assert(uuidPattern.test('not-a-uuid') === false)
         })
     })
+
+    describe('isUUID()', function () {
```
Comment (justinmk3) [inline packages/core/src/test/shared/crypto.test.ts:29]:
```suggestion
    describe('isUuid()', function () {
```

File: packages/core/src/shared/telemetry/util.ts [hunk 6]
```diff
@@ -90,6 +91,37 @@ export function convertLegacy(value: unknown): boolean {
     }
 }
 
+/**
+ * Returns an identifier that uniquely identifies a single application
+ * instance/window of a specific IDE. I.e if I have multiple VS Code
+ * windows open each one will have a unique session ID. This session ID
```
Comment (justinmk3) [inline packages/core/src/shared/telemetry/util.ts:97]: what matters is the toolkit instance, nothing else. we should be clear about that.
```suggestion
 * instance/window of a specific IDE. If I have multiple Toolkit extension instances (multiple vscode windows),
 * each one must have a unique session ID. This session ID
```

File: packages/core/src/shared/telemetry/util.ts [hunk 7]
```diff
@@ -90,6 +91,37 @@ export function convertLegacy(value: unknown): boolean {
...
+ * can be used in conjunction with the client ID to differntiate between
+ * different VS Code windows on a users machine.
+ *
+ * ### Rules:
+ *
+ * - On startup of a new application instance, a new session ID must be created.
+ *   - This identifier must be a `UUID`, it is enforced by the telemetry service.
+ * - The session ID must be different from all other IDs on the machine
+ * - A session ID exists until the application instance is terminated.
```
Comment (justinmk3) [inline packages/core/src/shared/telemetry/util.ts:106]:
```suggestion
 * - A session ID exists until the Toolkit extension instance is terminated.
```

File: packages/core/src/shared/telemetry/util.ts [hunk 8]
```diff
@@ -90,6 +91,37 @@ export function convertLegacy(value: unknown): boolean {
...
+ *   It should never be used again after termination.
+ * - All extensions on the same application instance should return the same session ID.
```
Comment (justinmk3) [inline packages/core/src/shared/telemetry/util.ts:108]: This might be an interesting, but is it actually needed? Perhaps this works by accident because of the way things work in vscode, but it might be difficult in other IDEs/systems? Does it matter either way (if not, probably shouldn't be a requirement)?

File: packages/core/src/test/shared/telemetry/util.test.ts [hunk 9]
```diff
@@ -107,6 +110,19 @@ describe('TelemetryConfig', function () {
...
+describe('getSessionId', function () {
+    it('returns a random UUIDv4', function () {
+        const result = getSessionId()
+
+        assert.deepStrictEqual(isUUID(result), true)
+        assert.deepStrictEqual(getSessionId(), result, 'Subsequent call did not return the same UUID')
+    })
+
+    it('nonce is the same as always', function () {
+        assert.deepStrictEqual(sessionIdNonce, '44cfdb20-b30b-4585-a66c-9f48f24f99b5')
```
Comment (justinmk3) [inline packages/core/src/test/shared/telemetry/util.test.ts:122]: do we need to test that a constant is constant?

File: packages/core/src/test/shared/telemetry/util.test.ts [hunk 10]
```diff
@@ -107,6 +110,19 @@ describe('TelemetryConfig', function () {
     })
 })
 
+describe('getSessionId', function () {
+    it('returns a random UUIDv4', function () {
```
Comment (justinmk3) [inline packages/core/src/test/shared/telemetry/util.test.ts:114]: test isn't checking randomness but it does check idempotency

```suggestion
    it('returns a stable UUIDv4', function () {
```

File: packages/core/src/shared/telemetry/util.ts (hunk 8 above)
Comment (nkomonen-amazon) [reply packages/core/src/shared/telemetry/util.ts:108]: I was thinking that if the Amazon Q and Toolkit extension are in the same window we want to know that since they interact with eachother (eg connection sharing).

This isn't enforced on the service side, so if other IDEs are not able to, then they can use different UUIDs. But then things won't be consistent across IDEs though.

Comment (justinmk3) [reply packages/core/src/shared/telemetry/util.ts:108]: > if the Amazon Q and Toolkit extension are in the same window we want to know

Could be useful. (Though it's an edge case since client-id will usually tell us that.)

File: packages/core/src/test/shared/telemetry/util.test.ts (hunk 10 above)
Comment (hayemaxi) [reply packages/core/src/test/shared/telemetry/util.test.ts:114]: Not a v4
```suggestion
    it('returns a stable UUID', function () {
```

File: packages/core/src/shared/crypto.ts [hunk 11]
```diff
@@ -30,3 +30,15 @@ export function randomUUID(): `${string}-${string}-${string}-${string}-${string}
...
+    const uuidv4Pattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
+    return uuidv4Pattern.test(uuid)
```
Comment (hayemaxi) [inline packages/core/src/shared/crypto.ts:43]:
```suggestion
    const uuidPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
    return uuidPattern.test(uuid)
```

File: packages/core/src/shared/telemetry/util.ts [hunk 12]
```diff
@@ -90,6 +91,37 @@ export function convertLegacy(value: unknown): boolean {
...
+ * instance/window of a specific IDE. I.e if I have multiple VS Code
+ * windows open each one will have a unique session ID. This session ID
+ * can be used in conjunction with the client ID to differntiate between
+ * different VS Code windows on a users machine.
+ *
+ * ### Rules:
+ *
+ * - On startup of a new application instance, a new session ID must be created.
+ *   - This identifier must be a `UUID`, it is enforced by the telemetry service.
+ * - The session ID must be different from all other IDs on the machine
```
[the remark this answers is not in the document]
Comment (nkomonen-amazon) [reply packages/core/src/shared/telemetry/util.ts:105]: This applies to Amazon Q as well as Toolkit. Are we fine using "Toolkit" as a broad name for one of our extensions?

Amazon Q and Toolkit on the same VSCode instance/window should have the same session id

File: packages/core/src/test/shared/telemetry/util.test.ts (hunk 9 above)
Comment (nkomonen-amazon) [reply packages/core/src/test/shared/telemetry/util.test.ts:122]: Any idea how this is possible? I couldn't find one

File: packages/core/src/shared/telemetry/util.ts (hunk 1 above)
Comment (nkomonen-amazon) [reply packages/core/src/shared/telemetry/util.ts:120]: Added this info to the comment.

File: packages/core/src/shared/crypto.ts (hunk 4 above)
Comment (nkomonen-amazon) [reply packages/core/src/shared/crypto.ts:41]: A stale comment I used to get Q to generate a pattern. Deleting this, but will add the new comment you suggested.

File: packages/core/src/shared/telemetry/util.ts (hunk 2 above)
Comment (hayemaxi) [reply packages/core/src/shared/telemetry/util.ts:111]: > We cannot guarantee the format of vscode.env.sessionId as it has no spec, at any point it could change to something else. Then the server would need to modify the pattern it accepts. UUID is a widely recognized format and is something we can consistently guarantee.

> The other IDEs will use this feature as well, if they have something similar to vscode.env.sessionId but its format is slightly different from ours, then should we modify the server to also accept that? I'd rather us have an obvious interface of requiring a UUID that all IDEs can easily follow.

Make the format any string and let the client determine what a "session ID" is, we should be accommodating to existing tools. The format validation isn't useful here - it doesn't stop clients from sending non-unique ids (and might make it harder to detect this if it "looks" right). 
Also, if there is a bug where clients send non-unique IDs, what is the outcome? It is not catastrophic like non-unique `clientId`

File: packages/core/src/shared/telemetry/util.ts (hunk 7 above)
Comment (nkomonen-amazon) [reply packages/core/src/shared/telemetry/util.ts:106]: If I have Amazon Q and Toolkit on my VS Code window, and Toolkit crashes, I want the session ID to remain the same for Amazon Q.

I'm not sure if this suggested change gets that across

File: packages/core/src/shared/telemetry/util.ts (hunk 8 above)
Comment (nkomonen-amazon) [reply packages/core/src/shared/telemetry/util.ts:108]: ClientID does not tell us this since it is the same across all VS Code windows on a machine. This was the original issue, if I had multiple VS Code windows with Amazon Q extension, and they were all sending telemetry, I could not differentiate between them. SessionID solves this.

**Review by hayemaxi — APPROVED** [review]
